sw/device/silicon_creator/rom_ext/rom_ext.c

```diff
@@ -189,16 +196,74 @@ static rom_error_t boot_svc_next_boot_bl0_slot_handler(
   boot_svc_next_boot_bl0_slot_res_init(kErrorOk,
                                        &boot_svc_msg->next_boot_bl0_slot_res);
 
-  HARDENED_RETURN_IF_ERROR(rom_ext_verify(kNextSlot));
+  HARDENED_RETURN_IF_ERROR(rom_ext_verify(kNextSlot, boot_data));
   // Boot fails if a verified ROM_EXT cannot be booted.
   HARDENED_RETURN_IF_ERROR(rom_ext_boot(kNextSlot));
   // `rom_ext_boot()` should never return `kErrorOk`, but if it does
   // we must shut down the chip instead of trying the next ROM_EXT.
   return kErrorRomExtBootFailed;
 }
 
+OT_WARN_UNUSED_RESULT
+static rom_error_t boot_svc_primary_boot_bl0_slot_handler(
+    boot_svc_msg_t *boot_svc_msg, boot_data_t *boot_data) {
+  uint32_t active_slot = boot_data->primary_bl0_slot;
+  uint32_t requested_slot = boot_svc_msg->primary_bl0_slot_req.primary_bl0_slot;
+
+  // In cases where the primary is already set to the requested slot, this
+  // function is a no-op.
+  if (active_slot != requested_slot) {
+    switch (launder32(requested_slot)) {
+      case kBootDataSlotA:
+        HARDENED_CHECK_EQ(requested_slot, kBootDataSlotA);
+        boot_data->primary_bl0_slot = kBootDataSlotA;
+        break;
+      case kBootDataSlotB:
+        HARDENED_CHECK_EQ(requested_slot, kBootDataSlotB);
+        boot_data->primary_bl0_slot = kBootDataSlotB;
+        break;
+      default:
+        HARDENED_TRAP();
+        OT_UNREACHABLE();
+    }
+
+    // Write boot data, updating relevant fields and recomputing the digest.
+    HARDENED_RETURN_IF_ERROR(boot_data_write(boot_data));
+    // Read the boot data back to ensure the correct slot is booted this time.
+    HARDENED_RETURN_IF_ERROR(boot_data_read(lc_state, boot_data));
+    HARDENED_RETURN_IF_ERROR(boot_data_check(boot_data));
+  }
+
+  boot_svc_primary_bl0_slot_res_init(boot_data->primary_bl0_slot, kErrorOk,
+                                     &boot_svc_msg->primary_bl0_slot_res);
+
+  return kErrorOk;
+}
+
+OT_WARN_UNUSED_RESULT
+static rom_error_t boot_svc_min_sec_ver_handler(boot_svc_msg_t *boot_svc_msg,
+                                                boot_data_t *boot_data) {
+  boot_data->min_security_version_bl0 =
+      boot_svc_msg->min_bl0_sec_ver_req.min_bl0_sec_ver;
```

Comment:
We must use a ratcheting increment here:
```suggestion
  const uint32_t current_min_sec_ver =  boot_data->min_security_version_bl0;
  const uint32_t requested_min_sec_ver = boot_svc_msg->min_bl0_sec_ver_req.min_bl0_sec_ver;
  const uint32_t diff = requested_min_sec_ver - current_min_sec_version;
  
  if (launder32(requested_min_sec_ver) >= current_sec_ver) {
    HARDENED_CHECK_GE(requested_min_sec_ver, current_sec_ver);
    HARDENED_CHECK_LE(diff, requested_min_sec_ver);
    ...
    // Proceed with the update.
    ...
  } else {
    return kErrorOk;
  }
```